crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1194,21 +1194,27 @@ impl<'db> Type<'db> {
                 ty.bool(db).is_always_true()
             }
 
-            (Type::SubclassOf(_), _) | (_, Type::SubclassOf(_)) => {
-                // TODO: Once we have support for final classes, we can determine disjointness in some cases
+            (Type::SubclassOf(_), other) | (other, Type::SubclassOf(_)) => {
+                // TODO: Once we have support for final classes, we can determine disjointness in more cases
                 // here. However, note that it might be better to turn `Type::SubclassOf('FinalClass')` into
                 // `Type::ClassLiteral('FinalClass')` during construction, instead of adding special cases for
                 // final classes inside `Type::SubclassOf` everywhere.
-                false
+                other.is_disjoint_from(db, KnownClass::Type.to_instance(db))
             }
 
-            (Type::KnownInstance(left), right) => {
-                left.instance_fallback(db).is_disjoint_from(db, right)
-            }
-            (left, Type::KnownInstance(right)) => {
-                left.is_disjoint_from(db, right.instance_fallback(db))
+            (Type::KnownInstance(known_instance), Type::Instance(InstanceType { class }))
+            | (Type::Instance(InstanceType { class }), Type::KnownInstance(known_instance)) => {
+                !known_instance.is_instance_of(db, class)
             }
 
+            (Type::KnownInstance(known_instance), Type::Tuple(_))
+            | (Type::Tuple(_), Type::KnownInstance(known_instance)) => KnownClass::Tuple
+                .to_class_literal(db)
+                .into_class_literal()
+                .is_some_and(|ClassLiteralType { class: tuple_class }| {
+                    !known_instance.class().is_subclass_of(db, tuple_class)
+                }),
```

Comment:
Would it be less repetitive to delegate this to "the known instance type is disjoint from `KnownClass::Tuple.to_instance()`, which would end up in the prior case?

Reply:
Ahh, great catch! I've pushed that change.